drivers/hwmon_ina219.c

```diff
@@ -0,0 +1,496 @@
+
+	return 1;
+}
+
+static int file_read_number(const char *path, int *value)
+{
+	char buf[128];
+	ssize_t ret;
+	int fd;
+
+	if ((fd = open(path, O_RDONLY)) < 0)
+		return -errno;
+
+	if ((ret = pread(fd, buf, sizeof(buf), 0)) < 0) {
+		const int _e = -errno;
+		close(fd);
+		return _e;
+	}
+	else {
+		close(fd);
+
+		buf[ret] = '\0';
+		*value = atoi(buf);
+		return 0;
+	}
+}
+
+static int check_ina219(const char *ina219_dir)
+{
+	char namepath[PATH_MAX];
+
+	upsdebugx(3, "checking %s", ina219_dir);
+
+	snprintf(namepath, sizeof(namepath), "%s/name", ina219_dir);
+
+	if (!file_contains(namepath, "ina219\n"))
+		return -ENODEV;
+
+	upsdebugx(3, "detected ina219 at %s", ina219_dir);
+	return 0;
+}
+
+static int detect_hwmon_ina219(const char *sysfs_hwmon_dir)
+{
+	DIR *sysfs;
+	struct dirent *entry;
+	int ret;
+
+	if (strcmp(device_path, "auto")) {
+		if ((ret = check_ina219(device_path)) < 0) {
+			fatal_with_errno(EXIT_FAILURE,
+					"not a valid hwmon ina219 dir: '%s'\n", device_path);
+		}
+
+		snprintf(ina219_base_path, sizeof(ina219_base_path), "%s", device_path);
+		return 0;
+	}
+
+	upslogx(LOG_NOTICE, "scanning %s for ina219", sysfs_hwmon_dir);
+
+	if ((sysfs = opendir(sysfs_hwmon_dir)) == NULL) {
+		const int _e = -errno;
+		upslog_with_errno(LOG_ERR, "unable to open %s", sysfs_hwmon_dir);
+		return _e;
+	}
+
+	while ((entry = readdir(sysfs)) != NULL) {
+		char hwmon_dir[PATH_MAX];
+
+		if (entry->d_type != DT_DIR && entry->d_type != DT_LNK) {
+			upsdebugx(3, "path %s/%s is not directory/symlink", sysfs_hwmon_dir,
+					entry->d_name);
+			continue;
+		}
+
+		if (!strcmp(entry->d_name, ".") || !strcmp(entry->d_name, "..")) {
+			upsdebugx(3, "skipping path %s/%s", sysfs_hwmon_dir, entry->d_name);
+			continue;
+		}
+
+		snprintf(hwmon_dir, sizeof(hwmon_dir), "%s/%s",
+					sysfs_hwmon_dir, entry->d_name);
+		if (check_ina219(hwmon_dir))
+			continue;
+
+		snprintf(ina219_base_path, sizeof(ina219_base_path), "%s", hwmon_dir);
+		closedir(sysfs);
+		return 0;
+	}
+
+	closedir(sysfs);
+	return -ENODEV;
+}
+
+static int update_intvar(
+	const char *base_path,
+	const char *name,
+	int *value)
+{
+	char path[PATH_MAX];
+	int ret;
+
+	if (snprintf(path, sizeof(path), "%s/%s", base_path, name) >= PATH_MAX) {
+		errno = ENAMETOOLONG;
+		upslog_with_errno(LOG_ERR, "snprintf(%s/%s) has failed", base_path, name);
+		return -ENAMETOOLONG;
+	}
+
+	if ((ret = file_read_number(path, value)) < 0) {
+		errno = -ret;
+		upslog_with_errno(LOG_ERR, "file_read_number(%s) has failed", path);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int update_voltage(void)
+{
+	return update_intvar(ina219_base_path, "in1_input", &voltage);
+}
+
+static int update_current(void)
+{
+	return update_intvar(ina219_base_path, "curr1_input", &current);
+}
+
+void upsdrv_makevartable(void)
+{
+	addvar(VAR_VALUE, "sysfs_dir",
+			"Path to sysfs dir of hwmon (" SYSFS_HWMON_DIR ")");
+	addvar(VAR_VALUE, "lb",
+			"Low battery threshold (" STRFY(BATTERY_CHARGE_LOW) ")");
```

Comment:
Note the comment above about `default.battery.charge.low` maybe being more maintainable, less surprising vs. other drivers.

Reply:
Changed.